<a href="https://colab.research.google.com/github/amirullahadisetya/Pemrograman-Dasar/blob/main/Praktikum%20Pertemuan%2010.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Program ini berfungsi untuk mengimpor empat library penting yang biasanya dipakai dalam analisis data. NumPy digunakan untuk membantu perhitungan angka dan operasi matematika, Pandas dipakai untuk mengolah data berbentuk tabel seperti layaknya Excel, Matplotlib digunakan untuk membuat berbagai jenis grafik, dan Seaborn dipakai untuk membuat grafik yang tampilannya lebih rapi dan menarik. Keempat baris ini intinya sedang menyiapkan alat-alat utama yang
dibutuhkan untuk analisis dan visualisasi data dalam Python.

In [ ]:
customers_df =pd.read_csv("/content/drive/MyDrive/PEMEROGRAMAN DASAR/pertemuan 10/Data/customers.csv")
customers_df.head()

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
0,1,fulan 1,Female,30,8606 Victoria TerraceSuite 560,5464,Johnstonhaven,Northern Territory,Australia
1,2,fulan 2,Prefer not to say,69,8327 Kirlin SummitApt. 461,8223,New Zacharyfort,South Australia,Australia
2,3,fulan 3,Prefer not to say,59,269 Gemma SummitSuite 109,5661,Aliburgh,Australian Capital Territory,Australia
3,4,fulan 4,Prefer not to say,67,743 Bailey GroveSuite 141,1729,South Justinhaven,Queensland,Australia
4,5,fulan 5,Prefer not to say,30,48 Hyatt ManorSuite 375,4032,Griffithsshire,Queensland,Australia


Kode tersebut membaca file customers.csv lalu menampilkan lima baris pertama dari data tersebut. Hasilnya menunjukkan bahwa dataset ini berisi informasi pelanggan, seperti ID, nama, jenis kelamin, usia, alamat, kode pos, kota, negara bagian, dan negara. Pada lima baris awal terlihat bahwa semua pelanggan berasal dari Australia tetapi tinggal di kota dan negara bagian yang berbeda-beda. Setiap baris merepresentasikan satu orang pelanggan lengkap dengan data demografinya, sehingga kode ini pada dasarnya digunakan untuk memastikan bahwa file berhasil dibaca dan untuk melihat gambaran awal isi data pelanggan tersebut.

In [ ]:
orders_df =pd.read_csv("/content/drive/MyDrive/PEMEROGRAMAN DASAR/pertemuan 10/Data/orders.csv")
orders_df.head()

,order_id,customer_id,payment,order_date,delivery_date
0,1,64,30811,2021-8-30,2021-09-24
1,2,473,50490,2021-2-3,2021-02-13
2,3,774,46763,2021-10-8,2021-11-03
3,4,433,39782,2021-5-6,2021-05-19
4,5,441,14719,2021-3-23,2021-03-24


Dataset orders_df berhasil dimuat dan menampilkan lima data pertama yang berisi informasi pesanan, seperti ID pesanan, ID pelanggan, jumlah pembayaran, tanggal pemesanan, dan tanggal pengiriman. Data terlihat lengkap, meskipun format tanggal pada kolom order_date masih belum konsisten. Secara keseluruhan, data ini menggambarkan proses pesanan dari dibuat hingga dikirim.

In [ ]:
products_df =pd.read_csv("/content/drive/MyDrive/PEMEROGRAMAN DASAR/pertemuan 10/Data/products.csv")
products_df.head()

,product_id,product_type,product_name,size,colour,price,quantity,description
0,0,Shirt,Oxford Cloth,XS,red,114,66,"A red coloured, XS sized, Oxford Cloth Shirt"
1,1,Shirt,Oxford Cloth,S,red,114,53,"A red coloured, S sized, Oxford Cloth Shirt"
2,2,Shirt,Oxford Cloth,M,red,114,54,"A red coloured, M sized, Oxford Cloth Shirt"
3,3,Shirt,Oxford Cloth,L,red,114,69,"A red coloured, L sized, Oxford Cloth Shirt"
4,4,Shirt,Oxford Cloth,XL,red,114,47,"A red coloured, XL sized, Oxford Cloth Shirt"


Kode tersebut membaca file products.csv dan menampilkan lima baris pertama dari data produk. Hasilnya menunjukkan bahwa dataset ini berisi informasi detail tentang produk yang dijual. Setiap baris mewakili satu produk dengan beberapa kolom penting, seperti product_id (ID produk), product_type (jenis produk), product_name (nama produk), size (ukuran), colour (warna), price (harga), quantity (jumlah stok), dan description (deskripsi produk). Dari lima baris awal terlihat bahwa semua produk yang muncul adalah kemeja “Oxford Cloth” dengan berbagai ukuran dari XS hingga XL, semuanya berwarna merah, dan memiliki stok serta harga yang berbeda-beda. Secara keseluruhan, kode ini digunakan untuk memastikan data produk berhasil dimuat serta untuk mengetahui isi awal katalog produk pada dataset tersebut.

In [ ]:
sales_df =pd.read_csv("/content/drive/MyDrive/PEMEROGRAMAN DASAR/pertemuan 10/Data/sales.csv")
sales_df.head()

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
0,0,1,218,106,2,212.0
1,1,1,481,118,1,118.0
2,2,1,2,96,3,288.0
3,3,1,1002,106,2,212.0
4,4,1,691,113,3,339.0


Kode ini membaca file sales.csv lalu menampilkan lima baris pertama datanya. Isi tabel menunjukkan detail penjualan, seperti ID penjualan, ID pesanan, ID produk, harga per unit, jumlah barang yang dibeli, dan total harga. Setiap baris menggambarkan satu item yang terjual dalam sebuah pesanan, sehingga kode ini membantu memastikan data penjualan berhasil dimuat dan memberikan gambaran awal tentang transaksi yang terjadi.

In [ ]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


Output tersebut adalah ringkasan struktur DataFrame customers_df. Artinya, data pelanggan memiliki 1007 baris dan 9 kolom. Setiap kolom berisi tipe data tertentu, seperti angka (int64) untuk customer_id, age, dan zip_code, serta teks (object) untuk kolom seperti customer_name, city, dan state. Kolom gender memiliki sedikit data kosong karena hanya ada 989 nilai yang terisi. Informasi ini membantu memahami ukuran dataset, tipe datanya, dan apakah ada kolom yang memiliki nilai hilang.

In [ ]:
customers_df.isna().sum()

,0
customer_id,0
customer_name,0
gender,18
age,0
home_address,0
zip_code,0
city,0
state,0
country,0


Hasil tersebut menunjukkan jumlah data yang hilang (NA) di setiap kolom. Semua kolom terisi lengkap kecuali kolom gender, yang memiliki 18 data kosong. Artinya, dari 1007 data pelanggan, ada 18 orang yang tidak mengisi atau tidak memiliki informasi jenis kelamin. Informasi ini berguna untuk mengetahui apakah perlu dilakukan pembersihan data, seperti mengisi nilai kosong atau menghapus baris tertentu.

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  6


Hasil tersebut menunjukkan bahwa dalam dataset customers_df terdapat 6 baris data yang terduplikasi. Artinya, ada 6 data pelanggan yang muncul lebih dari satu kali dengan isi baris yang sama persis. Informasi ini penting untuk proses pembersihan data, karena duplikasi bisa menyebabkan analisis menjadi tidak akurat jika tidak ditangani.

In [ ]:
customers_df.describe()

,customer_id,age,zip_code
count,1007.000000,1007.000000,1007.000000
mean,501.726912,50.929494,5012.538232
std,288.673238,30.516299,2885.836112
min,1.000000,20.000000,2.000000
25%,252.500000,34.000000,2403.500000
50%,502.000000,50.000000,5087.000000
75%,751.500000,65.000000,7493.500000
max,1000.000000,700.000000,9998.000000


Hasil customers_df.describe() memberikan ringkasan statistik untuk kolom angka dalam dataset pelanggan. Dari total 1007 data, terlihat bahwa nilai pada kolom customer_id dan zip_code tersebar dari angka kecil hingga mendekati 1000. Untuk kolom usia, rata-rata pelanggan berusia sekitar 50,9 tahun dengan rentang usia mulai dari 20 tahun hingga nilai ekstrem 700 tahun, yang kemungkinan merupakan data tidak realistis atau kesalahan input. Nilai kuartil menunjukkan bahwa sebagian besar pelanggan berada pada rentang usia 34 hingga 65 tahun. Secara keseluruhan, ringkasan ini membantu memahami distribusi data numerik, mengetahui kecenderungan umum, dan mendeteksi adanya nilai yang tidak wajar.

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


Output orders_df.info() menunjukkan bahwa data pesanan memiliki 1000 baris dan 5 kolom, yaitu order_id, customer_id, payment, order_date, dan delivery_date. Semua kolom terisi lengkap tanpa nilai kosong. Tiga kolom pertama bertipe angka (int64), sedangkan order_date dan delivery_date masih bertipe object karena disimpan dalam bentuk teks, bukan tanggal asli. Informasi ini membantu memastikan bahwa struktur data sudah benar, sekaligus menunjukkan bahwa kolom tanggal mungkin perlu dikonversi ke format datetime jika ingin dianalisis lebih lanjut.

In [ ]:
print("Jumlah duplikasi: ",orders_df.duplicated().sum())

Jumlah duplikasi:  0


Hasil tersebut menunjukkan bahwa tidak ada data yang terduplikasi di dalam tabel orders_df. Artinya, semua baris pesanan bersifat unik dan tidak ada catatan pesanan yang muncul lebih dari satu kali. Ini menandakan bahwa data pesanan bersih dari duplikasi dan siap digunakan untuk analisis lebih lanjut.

In [ ]:
orders_df.describe()

,order_id,customer_id,payment
count,1000.000000,1000.000000,1000.000000
mean,500.500000,506.640000,33972.936000
std,288.819436,277.115502,14451.609047
min,1.000000,1.000000,10043.000000
25%,250.750000,275.250000,21329.250000
50%,500.500000,515.000000,33697.500000
75%,750.250000,737.250000,46249.000000
max,1000.000000,1000.000000,59910.000000


Hasil orders_df.describe() menunjukkan bahwa data pesanan berisi 1000 transaksi dengan sebaran order_id dan customer_id yang merata dari 1 hingga 1000. Nilai pembayaran memiliki rata-rata sekitar 33.972, dengan pembayaran terkecil sekitar 10.043 dan terbesar hampir 59.910. Rentang kuartilnya menunjukkan bahwa sebagian besar pembayaran berada antara 21.329 hingga 46.249. Ringkasan ini memberikan gambaran singkat tentang variasi dan penyebaran nilai pembayaran dalam dataset.

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.3+ KB


Output products_df.info() menunjukkan bahwa dataset produk memiliki 1266 baris dan 8 kolom. Semua kolom terisi lengkap tanpa data kosong. Tiga kolom—product_id, price, dan quantity—bertipe angka, sedangkan kolom lainnya berisi teks seperti tipe produk, nama produk, ukuran, warna, dan deskripsi. Struktur ini memperlihatkan bahwa data produk tersusun rapi dan siap digunakan untuk analisis lebih lanjut tanpa perlu penanganan nilai hilang.

In [ ]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  6


Hasil tersebut menunjukkan bahwa pada dataset products_df terdapat 6 baris data yang terduplikasi. Artinya, ada enam produk yang muncul lebih dari satu kali dengan isi baris yang sama persis. Informasi ini penting karena duplikasi dapat memengaruhi analisis stok, harga, atau perhitungan penjualan jika tidak dibersihkan terlebih dahulu.

In [ ]:
products_df.describe()

,product_id,price,quantity
count,1266.000000,1266.000000,1266.000000
mean,627.926540,105.812006,60.138231
std,363.971586,9.715611,11.682791
min,0.000000,90.000000,40.000000
25%,313.250000,95.250000,50.000000
50%,626.500000,109.000000,60.000000
75%,942.750000,114.000000,70.000000
max,1259.000000,119.000000,80.000000


Hasil products_df.describe() memberikan ringkasan statistik untuk kolom angka dalam data produk. Dari 1266 produk, terlihat bahwa product_id tersebar dari 0 hingga 1259. Harga produk memiliki rata-rata sekitar 105, dengan harga termurah 90 dan termahal 119, menunjukkan variasi harga yang tidak terlalu besar. Kolom quantity (stok) rata-rata berada di angka 60, dengan stok terendah 40 dan tertinggi 80. Nilai kuartil menunjukkan bahwa sebagian besar produk memiliki stok antara 50 hingga 70. Secara singkat, ringkasan ini memberikan gambaran umum tentang pola harga dan jumlah stok produk dalam dataset.

In [ ]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


Output sales_df.info() menunjukkan bahwa data penjualan memiliki 5000 baris dan 6 kolom. Hampir semua kolom terisi lengkap, tetapi kolom total_price memiliki 19 data yang kosong, sementara kolom lainnya terisi penuh. Lima kolom bertipe angka bulat (int64), sedangkan total_price bertipe float karena berisi nilai desimal. Informasi ini membantu mengetahui bahwa sebagian besar data penjualan bersih, namun ada sedikit nilai hilang yang mungkin perlu diperbaiki sebelum analisis lebih lanjut.

In [ ]:
sales_df.isna().sum()

,0
sales_id,0
order_id,0
product_id,0
price_per_unit,0
quantity,0
total_price,19


Hasil tersebut menunjukkan bahwa seluruh kolom pada sales_df terisi lengkap kecuali kolom total_price, yang memiliki 19 data kosong. Artinya, ada 19 transaksi penjualan yang tidak memiliki perhitungan total harga. Informasi ini penting sebagai langkah awal pembersihan data, karena nilai yang hilang pada total harga mungkin perlu dihitung ulang atau ditangani sebelum melakukan analisis penjualan.

In [ ]:
print("Jumlah duplikasi: ", sales_df.duplicated().sum())

Jumlah duplikasi:  0


Hasil pengecekan menggunakan sales_df.duplicated().sum() menunjukkan bahwa jumlah duplikasi dalam data adalah 0. Artinya, seluruh baris pada dataset sales_df bersifat unik dan tidak ada data yang tercatat lebih dari satu kali. Kondisi ini bagus karena menandakan bahwa data penjualan yang digunakan sudah bersih dari pengulangan, sehingga analisis yang dilakukan nantinya bisa lebih akurat tanpa gangguan dari data ganda.

In [ ]:
sales_df.describe()

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.00000,4981.000000
mean,2499.500000,503.038200,634.053200,103.501600,1.99240,206.307368
std,1443.520003,285.964418,363.255794,9.195004,0.80751,86.352449
min,0.000000,1.000000,1.000000,90.000000,1.00000,90.000000
25%,1249.750000,258.000000,323.000000,95.000000,1.00000,112.000000
50%,2499.500000,504.500000,635.000000,102.000000,2.00000,204.000000
75%,3749.250000,749.000000,951.000000,112.000000,3.00000,285.000000
max,4999.000000,999.000000,1259.000000,119.000000,3.00000,357.000000


Statistik deskriptif pada sales_df menunjukkan bahwa dataset ini berisi 5.000 transaksi penjualan. Kolom total_price sedikit berbeda karena hanya memiliki 4.981 data valid akibat adanya nilai yang hilang. Rata-rata harga per unit produk berada di kisaran 103,5, dengan jumlah barang yang dibeli per transaksi sekitar 2 unit. Nilai minimum dan maksimum pada setiap kolom terlihat normal dan konsisten: misalnya, price_per_unit berada di rentang 90–119, sedangkan jumlah barang (quantity) berkisar antara 1–3 unit. Secara keseluruhan, data penjualan tampak stabil dan tidak menunjukkan penyimpangan ekstrem.

In [22]:
customers_df.drop_duplicates(inplace=True)

Perintah customers_df.drop_duplicates(inplace=True) digunakan untuk menghapus baris yang duplikat di dalam dataset customers_df. Dengan opsi inplace=True, proses penghapusan dilakukan langsung pada dataset aslinya tanpa membuat salinan baru. Jadi setelah perintah ini dijalankan, semua baris yang terduplikasi akan hilang dan customers_df berisi data unik saja.

In [23]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


Setelah dilakukan pengecekan ulang menggunakan perintah customers_df.duplicated().sum(), hasilnya menunjukkan jumlah duplikasi adalah 0. Artinya, di dalam dataset customers_df sudah tidak ada lagi baris yang kembar atau terduplikasi. Ini menandakan bahwa proses pembersihan data sebelumnya berhasil dan dataset kini bersih dari data ganda.

In [24]:
customers_df[customers_df.gender.isna()]

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
38,39,fulan 39,NaN,80,7440 Cameron Estate DrSuite 628,4622,North Victoriachester,Northern Territory,Australia
167,168,fulan 168,NaN,27,2781 Berge MallSuite 452,1975,North Leoburgh,Western Australia,Australia
322,322,fulan 322,NaN,30,593 Becker CircleApt. 333,1640,Jacobiview,Western Australia,Australia
393,393,fulan 393,NaN,34,5158 Levi HillSuite 531,1474,Johnsburgh,Queensland,Australia
442,442,fulan 442,NaN,26,5157 Feil RoadApt. 633,7249,Port Chloe,New South Wales,Australia
722,720,fulan 720,NaN,40,31 Jordan ParadeApt. 400,1380,West Henry,South Australia,Australia
745,743,fulan 743,NaN,57,09 Christopher StreetSuite 967,6226,Lake Lukemouth,Western Australia,Australia
773,771,fulan 771,NaN,74,7367 Wright JunctionApt. 773,8882,Kuhntown,Victoria,Australia
798,795,fulan 795,NaN,49,487 Summer MewsApt. 874,1712,East Hayden,Australian Capital Territory,Australia
801,798,fulan 798,NaN,56,27 Aiden KnollApt. 875,6531,Port Sam,Australian Capital Territory,Australia


Dari hasil pengecekan customers_df[customers_df.gender.isna()], terlihat bahwa terdapat beberapa baris data pelanggan yang tidak memiliki informasi jenis kelamin (gender bernilai NaN). Jika dilihat, seluruh pelanggan yang datanya hilang ini tetap memiliki informasi lain yang lengkap, seperti umur, alamat, kota, hingga negara. Artinya, masalah utama hanya ada pada kolom gender, sementara atribut lainnya tidak bermasalah. Baris-baris ini nantinya bisa ditangani dengan beberapa cara, seperti mengisi gender berdasarkan pola tertentu, menggunakan label “Unknown”, atau membiarkannya kosong jika tidak terlalu berpengaruh pada analisis. Yang jelas, data ini menunjukkan bahwa masih ada sedikit data yang perlu dibereskan sebelum dilakukan analisis lebih lanjut.

In [25]:
customers_df.gender.value_counts()

,count
gender,
Prefer not to say,725
Male,143
Female,115


Hasil customers_df.gender.value_counts() menunjukkan bahwa sebagian besar pelanggan memilih opsi “Prefer not to say” sebagai jenis kelamin, yaitu sebanyak 725 orang. Sementara itu, pelanggan yang menyatakan dirinya Male berjumlah 143 orang, dan yang berjenis kelamin Female sebanyak 115 orang. Data ini menunjukkan bahwa lebih dari setengah pelanggan tidak ingin mengungkapkan gender mereka, sehingga analisis berbasis gender perlu dilakukan dengan hati-hati karena distribusinya tidak seimbang

In [26]:
customers_df.isna().sum()

,0
customer_id,0
customer_name,0
gender,18
age,0
home_address,0
zip_code,0
city,0
state,0
country,0


Dari hasil customers_df.isna().sum(), terlihat bahwa hanya kolom gender yang memiliki data kosong, yaitu sebanyak 18 nilai yang hilang. Sementara itu, seluruh kolom lainnya seperti customer_id, customer_name, age, home_address, zip_code, city, state, dan country tidak memiliki missing value sama sekali. Artinya, masalah ketidaklengkapan data hanya terjadi pada informasi jenis kelamin, sehingga langkah pembersihan data dapat difokuskan pada kolom tersebut saja.

In [27]:
customers_df[customers_df.age == customers_df.age.max()]

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
967,961,fulan 961,Prefer not to say,700,29 Farrell ParadeSuite 818,6528,New Joseph,South Australia,Australia


Hasil pengecekan data menunjukkan bahwa terdapat satu pelanggan dengan usia 700 tahun, yaitu pada baris dengan customer_id 961. Nilai ini jelas tidak masuk akal dan menunjukkan adanya anomali atau kesalahan input dalam dataset. Informasi lain seperti nama, alamat, dan gender terlihat normal, namun usia tersebut sangat jauh di luar batas realistis. Karena itu, data ini sebaiknya ditinjau ulang, dikoreksi, atau dikeluarkan dari analisis agar tidak mempengaruhi kesimpulan yang dihasilkan.

In [28]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)

/tmp/ipython-input-4147959128.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 70, inplace=True)


Peringatan tersebut muncul karena Pandas memberi tahu bahwa cara Anda mengganti nilai dengan inplace=True pada subset kolom mungkin akan tidak berfungsi lagi di versi mendatang. Ini juga terkait dengan risiko chained assignment, yaitu ketika Pandas tidak yakin apakah Anda sedang memodifikasi data asli atau salinan.

In [29]:
customers_df[customers_df.age == customers_df.age.max()]

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
215,216,fulan 216,Prefer not to say,500,038 Haley MewsApt. 810,3991,Bayertown,Northern Territory,Australia


nilai usia maksimum di dataset kini berubah menjadi 500 tahun, yang berada pada data pelanggan dengan customer_id 216. Hal ini menunjukkan bahwa perbaikan sebelumnya berhasil mengganti nilai usia yang tidak realistis, namun masih terdapat nilai lain yang tetap tidak masuk akal untuk data manusia. Keberadaan usia yang sangat tinggi seperti ini menandakan adanya anomali atau kesalahan input dalam dataset, sehingga perlu diperhatikan sebelum analisis lebih lanjut dilakukan. Untuk mendapatkan hasil analisis yang lebih akurat, data seperti ini sebaiknya diperbaiki atau disesuaikan dengan nilai yang lebih masuk akal sesuai konteks penelitian.

In [30]:
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

/tmp/ipython-input-606824494.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 50, inplace=True)


Peringatan FutureWarning muncul karena cara mengganti nilai dengan inplace=True pada kolom dianggap tidak aman oleh Pandas dan bisa tidak bekerja di versi mendatang. Meskipun perintah masih berjalan, disarankan menggunakan metode yang lebih aman, seperti mengganti nilai langsung melalui assignment, agar perubahan selalu diterapkan pada data asli tanpa risiko kesalahan di masa depan.

In [31]:
customers_df[customers_df.age == customers_df.age.max()]

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
34,35,fulan 35,Male,80,3168 Bartoletti CrescentSuite 878,593,Port Lucas,Queensland,Australia
38,39,fulan 39,NaN,80,7440 Cameron Estate DrSuite 628,4622,North Victoriachester,Northern Territory,Australia
142,143,fulan 143,Prefer not to say,80,87 Hahn Station StSuite 943,8254,New Ella,Victoria,Australia
154,155,fulan 155,Male,80,85 Charles MallSuite 424,7841,Port Sophia,New South Wales,Australia
170,171,fulan 171,Prefer not to say,80,16 Boyer WaySuite 018,6226,South Gabriel,Western Australia,Australia
174,175,fulan 175,Prefer not to say,80,57 David MallSuite 596,3129,Haneton,South Australia,Australia
181,182,fulan 182,Male,80,32 Thomas CrestSuite 753,2147,Archiefurt,Queensland,Australia
235,236,fulan 236,Prefer not to say,80,64 Phillips RunApt. 722,2752,East Sienna,New South Wales,Australia
424,424,fulan 424,Prefer not to say,80,03 Grant CrestSuite 228,5743,Lake Charliemouth,Western Australia,Australia
438,438,fulan 438,Prefer not to say,80,569 Alyssa IslandApt. 766,6893,Hunterland,Western Australia,Australia


Hasil filter menunjukkan bahwa terdapat banyak pelanggan dengan usia 80 tahun, yang menjadi nilai maksimum setelah data sebelumnya diperbaiki. Data ini menampilkan berbagai informasi seperti nama, gender, alamat, dan lokasi, namun menunjukkan bahwa usia 80 muncul berulang kali pada banyak baris. Hal ini mengindikasikan bahwa nilai usia yang tidak realistis sebelumnya telah diganti dengan nilai yang lebih masuk akal, namun tetap menghasilkan pola yang perlu diperhatikan karena terlalu banyak data memiliki usia yang sama. Meskipun demikian, data kini sudah lebih rapi dan dapat digunakan untuk analisis, dengan catatan bahwa distribusi usia mungkin tidak sepenuhnya mencerminkan kondisi sebenarnya karena proses perbaikan data menggunakan nilai pengganti yang sama.

In [32]:
customers_df.describe()

,customer_id,age,zip_code
count,1001.000000,1001.000000,1001.000000
mean,500.942058,49.874126,5000.693307
std,289.013599,17.644663,2886.084454
min,1.000000,20.000000,2.000000
25%,251.000000,34.000000,2398.000000
50%,501.000000,50.000000,5079.000000
75%,751.000000,65.000000,7454.000000
max,1000.000000,80.000000,9998.000000


Hasil customers_df.describe() menunjukkan bahwa setelah proses pembersihan data, dataset pelanggan kini berisi 1001 data dengan distribusi nilai yang lebih masuk akal. Kolom customer_id dan zip_code memiliki sebaran angka yang merata, sementara rata-rata usia pelanggan berada di sekitar 49,8 tahun, dengan usia terendah 20 tahun dan tertinggi 80 tahun, yang jauh lebih realistis dibanding sebelumnya. Nilai kuartil menggambarkan bahwa sebagian besar pelanggan berada pada rentang usia 34 hingga 65 tahun, menunjukkan distribusi usia yang cukup wajar. Secara keseluruhan, ringkasan statistik ini memperlihatkan bahwa data sudah lebih bersih dan dapat digunakan dengan lebih aman untuk analisis lebih lanjut.

In [33]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

Perintah tersebut digunakan untuk mengubah data pada kolom order_date dan delivery_date menjadi format tanggal yang sebenarnya (datetime) sehingga Python dapat mengenali dan mengolahnya sebagai nilai waktu, bukan hanya teks. Dengan mengonversi kedua kolom tersebut, kita bisa melakukan berbagai analisis terkait waktu, seperti menghitung lamanya pengiriman, mencari pesanan berdasarkan periode tertentu, atau membuat visualisasi berdasarkan tanggal. Proses ini penting karena data bertipe datetime memberi kemampuan analisis yang jauh lebih kuat dibanding data berupa teks biasa.

In [34]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


Hasil orders_df.info() menunjukkan bahwa dataset pesanan kini memiliki semua data yang lengkap tanpa nilai kosong, dengan total 1000 baris dan 5 kolom. Kolom order_date dan delivery_date sudah berhasil diubah menjadi tipe datetime, sehingga Python dapat menangani kedua kolom tersebut sebagai data waktu yang valid. Tiga kolom lainnya tetap berupa angka. Dengan perubahan ini, dataset menjadi lebih mudah dan lebih kuat untuk dianalisis, terutama jika ingin mengolah informasi terkait waktu, seperti menghitung lama pengiriman atau melihat pola pesanan berdasarkan tanggal.

In [35]:
products_df.drop_duplicates(inplace=True)

Perintah products_df.drop_duplicates(inplace=True) digunakan untuk menghapus semua baris data produk yang duplikat di dalam dataset. Dengan menggunakan inplace=True, penghapusan terjadi langsung pada dataset asli tanpa membuat salinan baru, sehingga setelah perintah ini dijalankan, data produk yang tersisa hanya yang benar-benar unik. Langkah ini penting karena duplikasi dapat membuat analisis seperti perhitungan stok atau penjualan menjadi tidak akurat, sehingga membersihkan data dari baris yang sama akan menghasilkan informasi yang lebih valid dan dapat dipercaya.

In [36]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  0


Hasil pengecekan menggunakan products_df.duplicated().sum() menunjukkan bahwa jumlah duplikasi dalam dataset produk adalah 0, yang berarti tidak ada lagi baris data yang kembar setelah proses pembersihan dilakukan. Kondisi ini menunjukkan bahwa dataset kini sudah bersih dan hanya berisi data produk yang unik, sehingga analisis selanjutnya dapat dilakukan dengan lebih akurat tanpa terpengaruh oleh data ganda.

In [37]:
sales_df[sales_df.total_price.isna()]

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
9,9,2,1196,105,1,NaN
121,121,27,1027,90,3,NaN
278,278,63,360,94,2,NaN
421,421,95,1091,115,1,NaN
489,489,108,1193,105,3,NaN
539,539,117,405,119,2,NaN
636,636,134,653,93,3,NaN
687,687,145,1138,102,1,NaN
854,854,177,64,104,1,NaN
1079,1079,222,908,94,3,NaN


Baris dengan total_price = NaN menunjukkan transaksi yang belum memiliki nilai total, meskipun price_per_unit dan quantity sudah lengkap. Nilai yang hilang ini bisa diperbaiki dengan menghitung ulang total_price melalui rumus price_per_unit × quantity, sehingga dataset menjadi rapi dan siap digunakan untuk analisis penjualan.

In [40]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

Kode tersebut digunakan untuk mengisi kolom total_price yang sebelumnya berisi nilai kosong (NaN). Dengan menghitung price_per_unit × quantity, setiap transaksi kini memiliki nilai total harga yang lengkap. Hasilnya, data pada sales_df menjadi lebih bersih, konsisten, dan siap dipakai untuk analisis lebih lanjut.

In [39]:
sales_df.isna().sum()

,0
sales_id,0
order_id,0
product_id,0
price_per_unit,0
quantity,0
total_price,0


Semua nilai NaN sudah hilang di sales_df. Setiap kolom menunjukkan jumlah nilai kosong 0, yang berarti dataset sudah bersih dan lengkap. Ini menandakan proses pembersihan, khususnya pengisian kolom total_price, berhasil sepenuhnya sehingga dataset siap digunakan untuk analisis berikutnya.

In [41]:
customers_df.describe(include="all")

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country
count,1001.000000,1001,983,1001.000000,1001,1001.000000,1001,1001,1001
unique,NaN,1000,3,NaN,1000,NaN,961,8,1
top,NaN,fulan 943,Prefer not to say,NaN,3117 Heller PlaceSuite 149,NaN,East Aidan,South Australia,Australia
freq,NaN,2,725,NaN,2,NaN,3,140,1001
mean,500.942058,NaN,NaN,49.874126,NaN,5000.693307,NaN,NaN,NaN
std,289.013599,NaN,NaN,17.644663,NaN,2886.084454,NaN,NaN,NaN
min,1.000000,NaN,NaN,20.000000,NaN,2.000000,NaN,NaN,NaN
25%,251.000000,NaN,NaN,34.000000,NaN,2398.000000,NaN,NaN,NaN
50%,501.000000,NaN,NaN,50.000000,NaN,5079.000000,NaN,NaN,NaN
75%,751.000000,NaN,NaN,65.000000,NaN,7454.000000,NaN,NaN,NaN


Hasil describe(include="all") pada customers_df menunjukkan bahwa data pelanggan cukup lengkap dan beragam. Kolom nama, alamat, kota, dan state memiliki banyak nilai unik, menandakan variasi tinggi. Mayoritas pelanggan memilih gender “Prefer not to say”. Data numerik seperti customer_id, age, dan zip_code juga memiliki sebaran yang wajar. Seluruh pelanggan berasal dari Australia. Secara keseluruhan, dataset terlihat bersih dan tidak bermasalah.

In [42]:
customers_df.groupby(by="gender").agg({
    "customer_id": "nunique",
    "age": ["max", "min", "mean", "std"]
})

customer_id age                          
                      nunique max min       mean        std
gender                                                     
Female                    115  79  20  49.147826  16.646607
Male                      143  80  20  51.230769  18.462635
Prefer not to say         725  80  20  49.748966  17.658998

Hasil groupby berdasarkan gender menunjukkan bahwa kelompok “Prefer not to say” memiliki jumlah pelanggan terbanyak, yaitu 725 orang, jauh lebih besar dibandingkan Female (115) dan Male (143). Dari sisi usia, ketiga kelompok memiliki rentang umur yang sama, yaitu antara 20 hingga sekitar 79–80 tahun. Rata-rata usia pelanggan juga tidak jauh berbeda, berada di kisaran 49–51 tahun, dengan variasi umur (standar deviasi) yang relatif mirip. Secara keseluruhan, distribusi usia antar gender cukup seimbang, hanya jumlah pelanggannya saja yang berbeda jauh.

In [43]:
customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

,customer_id
city,
East Sophia,3
New Ava,3
East Aidan,3
Lake Sienna,2
Jademouth,2
...,...
Justinport,1
Kaiberg,1
Kaifort,1


Hasil pengelompokan berdasarkan city menunjukkan bahwa sebagian besar kota hanya memiliki 1 pelanggan, terlihat dari total 961 kota yang muncul dalam data. Hanya beberapa kota yang memiliki lebih dari satu pelanggan, seperti East Sophia, New Ava, dan East Aidan, masing-masing dengan 3 pelanggan, serta beberapa kota lain seperti Lake Sienna dan Jademouth dengan 2 pelanggan. Secara keseluruhan, data ini menggambarkan bahwa persebaran pelanggan sangat tersebar dan tidak terpusat pada kota tertentu.

In [44]:
customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

,customer_id
state,
South Australia,139
Queensland,134
New South Wales,132
Northern Territory,125
Western Australia,124
Australian Capital Territory,121
Victoria,121
Tasmania,104


Dari hasil pengelompokan berdasarkan state, terlihat bahwa jumlah pelanggan tersebar cukup merata di seluruh wilayah Australia, namun dengan sedikit perbedaan dominasi. South Australia memiliki jumlah pelanggan terbanyak yaitu 139 orang, diikuti oleh Queensland (134) dan New South Wales (132). Sementara itu, wilayah seperti Tasmania memiliki jumlah pelanggan paling sedikit, yaitu 104 orang. Secara keseluruhan, distribusi pelanggan antar negara bagian cukup seimbang dan tidak ada perbedaan yang terlalu mencolok.

In [45]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)

Kode tersebut menghitung lama waktu pengiriman setiap pesanan dengan cara mengurangkan delivery_date dari order_date, lalu mengubah selisih waktu tersebut menjadi detik. Setelah itu, nilai detik dikonversi ke satuan hari dengan membaginya menggunakan 86.400 detik (jumlah detik dalam satu hari). Terakhir, hasilnya dibulatkan dan disimpan ke kolom baru bernama delivery_time pada orders_df. Dengan demikian, setiap baris kini memiliki informasi lama pengiriman dalam bentuk jumlah hari.

In [46]:
orders_df.describe(include="all")

,order_id,customer_id,payment,order_date,delivery_date,delivery_time
count,1000.000000,1000.000000,1000.000000,1000,1000,1000.000000
mean,500.500000,506.640000,33972.936000,2021-05-27 18:38:52.800000,2021-06-10 20:31:12,14.078000
min,1.000000,1.000000,10043.000000,2021-01-01 00:00:00,2021-01-03 00:00:00,1.000000
25%,250.750000,275.250000,21329.250000,2021-03-13 18:00:00,2021-03-28 00:00:00,8.000000
50%,500.500000,515.000000,33697.500000,2021-05-27 12:00:00,2021-06-11 12:00:00,14.000000
75%,750.250000,737.250000,46249.000000,2021-08-12 06:00:00,2021-08-24 06:00:00,21.000000
max,1000.000000,1000.000000,59910.000000,2021-10-24 00:00:00,2021-11-20 00:00:00,27.000000
std,288.819436,277.115502,14451.609047,NaN,NaN,7.707225


Hasil describe pada orders_df menunjukkan bahwa data berisi 1000 pesanan dengan rata-rata pembayaran sekitar 33.972. Tanggal pemesanan dan pengiriman berada sepanjang tahun 2021. Waktu pengiriman (delivery_time) rata-rata 14 hari, dengan rentang 1 hingga 27 hari. Secara umum, sebagian besar pesanan tiba dalam waktu 1–3 minggu.

In [47]:
customer_id_in_orders_df =  orders_df.customer_id.tolist()
customers_df["status"] = customers_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_orders_df else "Non Active")
customers_df.sample(5)

,customer_id,customer_name,gender,age,home_address,zip_code,city,state,country,status
51,52,fulan 52,Male,47,012 Cooper MallApt. 259,841,Chelseahaven,Victoria,Australia,Active
430,430,fulan 430,Female,56,4112 Lincoln BoulevardApt. 887,8212,Ivyfurt,New South Wales,Australia,Non Active
189,190,fulan 190,Prefer not to say,58,63 Rohan PlazaApt. 943,5518,Lake Charlottefort,Australian Capital Territory,Australia,Non Active
672,671,fulan 671,Prefer not to say,53,0831 Hahn HillApt. 652,587,East Abbeyborough,Tasmania,Australia,Non Active
414,414,fulan 414,Prefer not to say,51,22 Eve CrescentApt. 357,7365,North Ivy,Queensland,Australia,Active


Setelah menambahkan kolom status, setiap pelanggan pada customers_df diklasifikasikan sebagai Active atau Non Active berdasarkan apakah customer_id-nya muncul dalam data pesanan (orders_df). Contoh hasilnya menunjukkan bahwa beberapa pelanggan, seperti fulan 52 dan fulan 414, berstatus Active karena pernah melakukan transaksi. Sementara itu, pelanggan lain seperti fulan 430, fulan 190, dan fulan 671 berstatus Non Active, yang berarti mereka tidak tercatat melakukan pemesanan dalam dataset. Pendekatan ini membantu memisahkan pelanggan aktif dan pasif dengan jelas.

In [48]:
customers_df.groupby(by="status").customer_id.count()

,customer_id
status,
Active,617
Non Active,384


Hasil pengelompokan data berdasarkan status menunjukkan bahwa dari total 1001 pelanggan, sebanyak 617 pelanggan termasuk kategori Active, artinya mereka pernah melakukan transaksi. Sementara itu, 384 pelanggan lainnya masuk kategori Non Active karena tidak ditemukan pada data pesanan. Informasi ini memberikan gambaran bahwa mayoritas pelanggan dalam dataset adalah pengguna aktif, sehingga bisa menjadi fokus utama untuk analisis perilaku atau strategi pemasaran.

In [49]:
orders_customers_df = pd.merge(
    left=orders_df,
    right=customers_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id"
)
orders_customers_df.head()

,order_id,customer_id,payment,order_date,delivery_date,delivery_time,customer_name,gender,age,home_address,zip_code,city,state,country,status
0,1,64,30811,2021-08-30,2021-09-24,25.0,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active
1,2,473,50490,2021-02-03,2021-02-13,10.0,fulan 473,Male,61,531 Schmitt BoulevardApt. 010,1744,Annaton,South Australia,Australia,Active
2,3,774,46763,2021-10-08,2021-11-03,26.0,fulan 774,Prefer not to say,34,2096 Wilson MewsApt. 714,8590,West Jonathanshire,Tasmania,Australia,Active
3,4,433,39782,2021-05-06,2021-05-19,13.0,fulan 433,Prefer not to say,46,5777 Mayer PassApt. 881,9728,West Michaelport,Tasmania,Australia,Active
4,5,441,14719,2021-03-23,2021-03-24,1.0,fulan 441,Prefer not to say,53,33 Richards JunctionApt. 478,7650,South Rileyview,Western Australia,Australia,Active


Setelah dilakukan proses merge antara orders_df dan customers_df, terbentuklah dataframe baru bernama orders_customers_df yang menggabungkan informasi pesanan dan data pelanggan berdasarkan customer_id. Hasil penggabungan ini menunjukkan bahwa setiap baris pesanan kini dilengkapi dengan detail pelanggan seperti nama, gender, usia, alamat, hingga status aktif atau tidaknya pelanggan. Contoh pada lima data teratas terlihat bahwa seluruh pesanan berhasil terhubung dengan data pelanggan yang sesuai, sehingga dataset ini kini lebih lengkap dan siap digunakan untuk analisis lanjutan.

In [50]:
orders_customers_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

,city,order_id
0,Jordanside,6
1,New Ava,6
2,Port Hannahburgh,5
3,Lake Rose,5
4,O'keefeton,5
5,East Max,5
6,Rubyfort,5
7,West Kai,5
8,Sarahstad,4
9,Russelstad,4


Berdasarkan hasil pengelompokan data pada orders_customers_df, kota dengan jumlah pesanan terbanyak adalah Jordanside dan New Ava, masing-masing mencatat 6 pesanan unik. Menyusul di bawahnya terdapat beberapa kota seperti Port Hannahburgh, Lake Rose, O'keefeton, East Max, dan Rubyfort yang masing-masing memiliki 5 pesanan. Sementara itu, kota seperti West Kai, Sarahstad, dan Russelstad mencatat 4 pesanan. Informasi ini menunjukkan bahwa beberapa kota tertentu menjadi pusat aktivitas pembelian yang lebih tinggi dibandingkan kota lainnya.

In [51]:
orders_customers_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

,order_id
state,
South Australia,148
Queensland,139
Western Australia,130
New South Wales,129
Victoria,118
Australian Capital Territory,118
Tasmania,112
Northern Territory,106


Dari hasil pengelompokan berdasarkan state, terlihat bahwa South Australia menjadi wilayah dengan pesanan terbanyak, yaitu 148 pesanan unik. Posisi berikutnya diisi oleh Queensland dengan 139 pesanan, disusul Western Australia (130 pesanan) dan New South Wales (129 pesanan). Sementara itu, Victoria serta Australian Capital Territory masing-masing memiliki 118 pesanan, Tasmania mencatat 112 pesanan, dan Northern Territory menjadi yang paling sedikit dengan 106 pesanan. Data ini menunjukkan bahwa aktivitas transaksi paling tinggi terjadi di South Australia dan paling rendah di Northern Territory.

In [52]:
orders_customers_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)

,order_id
gender,
Prefer not to say,716
Female,139
Male,136


Dari hasil pengelompokan berdasarkan gender, terlihat bahwa pelanggan dengan kategori “Prefer not to say” melakukan pesanan terbanyak, yaitu 716 order. Sementara itu, pelanggan berjenis kelamin Female mencatat 139 order, dan Male sedikit lebih rendah dengan 136 order. Hal ini menunjukkan bahwa kelompok yang memilih tidak menyebutkan jenis kelamin mendominasi jumlah transaksi secara signifikan dibanding dua kategori lainnya.

In [53]:
orders_customers_df["age_group"] = orders_customers_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))

Kode tersebut menambahkan kolom age_group pada orders_customers_df dengan mengelompokkan usia pelanggan menjadi tiga kategori: Youth (≤24 tahun), Adults (25–64 tahun), dan Seniors (>64 tahun). Pengelompokan ini membantu memudahkan analisis berdasarkan kelompok usia.

In [54]:
orders_customers_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)

,order_id
age_group,
Adults,681
Seniors,226
Youth,93


Hasil pengelompokan order berdasarkan age_group menunjukkan bahwa kelompok Adults melakukan pemesanan paling banyak dengan 681 order, diikuti oleh Seniors sebanyak 226 order, dan yang paling sedikit adalah kelompok Youth dengan 93 order. Ini menunjukkan bahwa pelanggan usia dewasa merupakan kontributor utama dalam jumlah transaksi.

In [55]:
products_df.describe(include="all")

,product_id,product_type,product_name,size,colour,price,quantity,description
count,1260.000000,1260,1260,1260,1260,1260.000000,1260.000000,1260
unique,NaN,3,35,8,7,NaN,NaN,1260
top,NaN,Shirt,Denim,XS,red,NaN,NaN,"A violet coloured, XL sized, Tracksuit Bottoms..."
freq,NaN,420,70,252,180,NaN,NaN,1
mean,629.500000,NaN,NaN,NaN,NaN,105.805556,60.150000,NaN
std,363.874979,NaN,NaN,NaN,NaN,9.704423,11.670573,NaN
min,0.000000,NaN,NaN,NaN,NaN,90.000000,40.000000,NaN
25%,314.750000,NaN,NaN,NaN,NaN,95.750000,50.000000,NaN
50%,629.500000,NaN,NaN,NaN,NaN,108.500000,60.000000,NaN
75%,944.250000,NaN,NaN,NaN,NaN,114.000000,70.000000,NaN


Hasil describe pada products_df menunjukkan bahwa dataset berisi 1.260 produk dengan tiga jenis utama, yaitu Shirt, Pants, dan Tracksuit, di mana jenis Shirt menjadi yang paling banyak muncul. Terdapat 35 nama produk, 8 ukuran, dan 7 warna yang berbeda. Dari sisi numerik, harga produk berada pada rentang 90 hingga 119, dengan harga rata-rata sekitar 105.8. Sementara itu, jumlah stok produk berkisar 40 hingga 80 unit dengan rata-rata sekitar 60 unit. Setiap deskripsi produk bersifat unik, karena seluruh baris memiliki deskripsi yang berbeda. Secara keseluruhan, data produk cukup beragam baik dari segi tipe, ukuran, warna, maupun stok.

In [56]:
sales_df.describe(include="all")

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.00000,5000.000000
mean,2499.500000,503.038200,634.053200,103.501600,1.99240,206.360000
std,1443.520003,285.964418,363.255794,9.195004,0.80751,86.357457
min,0.000000,1.000000,1.000000,90.000000,1.00000,90.000000
25%,1249.750000,258.000000,323.000000,95.000000,1.00000,112.000000
50%,2499.500000,504.500000,635.000000,102.000000,2.00000,204.000000
75%,3749.250000,749.000000,951.000000,112.000000,3.00000,285.000000
max,4999.000000,999.000000,1259.000000,119.000000,3.00000,357.000000


Hasil describe pada sales_df menunjukkan bahwa dataset terdiri dari 5.000 transaksi penjualan. Harga per unit produk berada pada rentang 90 hingga 119, dengan rata-rata sekitar 103.5. Jumlah barang yang dibeli per transaksi berkisar antara 1 hingga 3 unit, dan nilai total transaksi (total_price) berada antara 90 hingga 357, dengan rata-rata sekitar 206. Data ini juga menunjukkan bahwa distribusi order_id dan product_id tersebar merata tanpa nilai yang hilang, sehingga keseluruhan dataset terlihat bersih dan konsisten untuk dianalisis lebih lanjut.

In [57]:
products_df.sort_values(by="price", ascending=False)

,product_id,product_type,product_name,size,colour,price,quantity,description
671,665,Jacket,Parka,XS,red,119,65,"A red coloured, XS sized, Parka Jacket"
672,666,Jacket,Parka,S,red,119,66,"A red coloured, S sized, Parka Jacket"
673,667,Jacket,Parka,M,red,119,56,"A red coloured, M sized, Parka Jacket"
674,668,Jacket,Parka,L,red,119,73,"A red coloured, L sized, Parka Jacket"
675,669,Jacket,Parka,XL,red,119,64,"A red coloured, XL sized, Parka Jacket"
...,...,...,...,...,...,...,...,...
1222,1216,Trousers,Pleated,S,indigo,90,41,"A indigo coloured, S sized, Pleated Trousers"
1229,1223,Trousers,Pleated,L,violet,90,48,"A violet coloured, L sized, Pleated Trousers"
1228,1222,Trousers,Pleated,M,violet,90,45,"A violet coloured, M sized, Pleated Trousers"
1227,1221,Trousers,Pleated,S,violet,90,50,"A violet coloured, S sized, Pleated Trousers"


Ketika products_df diurutkan berdasarkan harga tertinggi, terlihat bahwa produk dengan harga 119 merupakan yang paling mahal, seluruhnya berasal dari kategori Jacket tipe Parka dengan variasi ukuran XS hingga XL dan warna merah. Setelah itu, harga produk menurun secara bertahap hingga harga terendah yaitu 90, yang umumnya berasal dari kategori Trousers tipe Pleated. Urutan ini menunjukkan bahwa kategori Jacket memiliki nilai jual tertinggi dalam katalog, sementara produk celana (Trousers) berada pada kisaran harga yang paling rendah.

In [58]:
products_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price":  ["min", "max"]
})

product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119

Pengelompokan berdasarkan product_type menunjukkan bahwa Jacket, Shirt, dan Trousers masing-masing memiliki 420 produk unik dengan jumlah stok yang hampir sama. Harga ketiga kategori juga konsisten, berkisar antara 90 hingga 119. Data ini menggambarkan bahwa perusahaan menjaga keseragaman dalam variasi produk, stok, dan rentang harga di setiap jenis produk.

In [59]:
products_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})


product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              35     2017    93   93
Denim                     70     4204    92  115
Drawstring                35     2042   104  104
Dress                     35     2125   119  119
Flannel                   35     2135    96   96
Henley                    35     2051    92   92
High-Waisted              35     2198    98   98
Joggers                   35     2107    94   94
Leather                   35     2276   113  113
Linen                     35     2138   116  116
Mandarin Collar           35     2203   108  108
Oxford Cloth              35     2071   114  114
Parka                     35     2201   119  119
Peacoat                   35     2067   102  102
Pleated                   35     2030    90   90
Polo                      35     2196   117  117
Puffer                    35     1946   110  110
Pullover                  35     2184   114  114
Relaxed Leg               35     2002    95   95
Shearling                 35     2169    95   95
Slim-Fit                  35     2155   119  119
Tracksuit Bottoms         35     2038    91   91
Trench Coat               35     2132   112  112
Windbreaker               35     2085   109  109
Wool                      35     1975   111  111

Data tersebut menunjukkan bahwa setiap product_name memiliki 35 produk unik (kecuali Denim yang memiliki 70) dengan total stok yang relatif mirip antar kategori. Harga untuk tiap produk juga konsisten, karena setiap nama produk hanya memiliki satu rentang harga tetap—misalnya Bomber selalu berharga 90, Dress selalu 119, dan Linen selalu 116. Denim menjadi satu-satunya kategori dengan variasi harga (92 hingga 115) karena memiliki jumlah produk lebih banyak. Secara keseluruhan, perusahaan menjaga konsistensi harga dan distribusi stok di hampir semua jenis produk, dengan Denim sebagai pengecualian yang menawarkan variasi lebih luas.

In [60]:
sales_products_df = pd.merge(
    left=sales_df,
    right=products_df,
    how="left",
    left_on="product_id",
    right_on="product_id"
)
sales_products_df.head()

,sales_id,order_id,product_id,price_per_unit,quantity_x,total_price,product_type,product_name,size,colour,price,quantity_y,description
0,0,1,218,106,2,212,Shirt,Chambray,L,orange,105,44,"A orange coloured, L sized, Chambray Shirt"
1,1,1,481,118,1,118,Jacket,Puffer,S,indigo,110,62,"A indigo coloured, S sized, Puffer Jacket"
2,2,1,2,96,3,288,Shirt,Oxford Cloth,M,red,114,54,"A red coloured, M sized, Oxford Cloth Shirt"
3,3,1,1002,106,2,212,Trousers,Wool,M,blue,111,52,"A blue coloured, M sized, Wool Trousers"
4,4,1,691,113,3,339,Jacket,Parka,S,indigo,119,53,"A indigo coloured, S sized, Parka Jacket"


Hasil merge antara sales_df dan products_df menghasilkan sales_products_df, yaitu tabel penjualan yang sudah dilengkapi dengan informasi detail produk. Setiap baris kini tidak hanya berisi data transaksi seperti sales_id, order_id, jumlah pembelian, dan total harga, tetapi juga data produk lengkap seperti product_type, product_name, ukuran, warna, harga asli produk, hingga deskripsi barang. Dengan tabel ini, analisis penjualan menjadi lebih mudah karena data transaksi dan karakteristik produk sudah tergabung dalam satu dataset yang terpadu.

In [61]:
sales_products_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

,sales_id,quantity_x,total_price
product_type,,,
Jacket,1676,3343,357026
Shirt,1641,3259,333600
Trousers,1683,3360,341174


Dari hasil pengelompokan berdasarkan product_type, terlihat bahwa ketiga kategori—Jacket, Shirt, dan Trousers—memiliki jumlah transaksi yang hampir seimbang. Namun, Trousers mencatat jumlah penjualan tertinggi dengan 1.683 transaksi dan total kuantitas 3.360 unit, menghasilkan pendapatan sekitar 341.174. Jacket menyusul dengan pendapatan tertinggi yaitu 357.026 dari 1.676 transaksi, sementara Shirt mencatat pendapatan 333.600 dari 1.641 transaksi. Secara umum, ketiga kategori menunjukkan performa penjualan yang stabil dengan kontribusi pendapatan yang cukup merata.

In [62]:
sales_products_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
}).sort_values(by="total_price", ascending=False)

,sales_id,quantity_x,total_price
product_name,,,
Denim,273,527,52399
Casual Slim Fit,154,306,36414
Trench Coat,146,299,35581
Shearling,150,302,35334
Puffer,140,298,35164
Flannel,141,281,33158
Cropped,135,284,32660
Pleated,147,308,32340
Joggers,164,334,31062


Hasil pengelompokan berdasarkan product_name menunjukkan bahwa Denim menjadi produk dengan performa penjualan terbaik, mencatat total pendapatan tertinggi sebesar 52.399 dari 273 transaksi dan 527 unit terjual. Di bawahnya terdapat produk seperti Casual Slim Fit, Trench Coat, Shearling, dan Puffer yang masing-masing menghasilkan pendapatan antara 35.000 hingga 36.000, menunjukkan minat konsumen yang juga tinggi. Sementara itu, produk seperti Cords, Linen, dan Coach berada pada posisi terbawah dalam daftar pendapatan, meskipun tetap memberikan kontribusi yang stabil. Secara keseluruhan, perbedaan total pendapatan ini mencerminkan variasi popularitas tiap produk, dengan Denim sebagai kategori yang paling diminati dalam penjualan.

In [63]:
all_df = pd.merge(
    left=sales_products_df,
    right=orders_customers_df,
    how="left",
    left_on="order_id",
    right_on="order_id"
)
all_df.head()

,sales_id,order_id,product_id,price_per_unit,quantity_x,total_price,product_type,product_name,size,colour,...,customer_name,gender,age,home_address,zip_code,city,state,country,status,age_group
0,0,1,218,106,2,212,Shirt,Chambray,L,orange,...,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active,Seniors
1,1,1,481,118,1,118,Jacket,Puffer,S,indigo,...,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active,Seniors
2,2,1,2,96,3,288,Shirt,Oxford Cloth,M,red,...,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active,Seniors
3,3,1,1002,106,2,212,Trousers,Wool,M,blue,...,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active,Seniors
4,4,1,691,113,3,339,Jacket,Parka,S,indigo,...,fulan 64,Prefer not to say,75,4927 Alice MeadowApt. 960,7787,Sanfordborough,South Australia,Australia,Active,Seniors


Dataframe all_df adalah hasil gabungan lengkap antara data penjualan, produk, pesanan, dan pelanggan sehingga menghasilkan 28 kolom informasi. Tabel ini berisi detail transaksi sekaligus profil pelanggan—mulai dari jenis produk, jumlah pembelian, nilai total, usia, gender, hingga kota dan status keanggotaannya. Dengan struktur ini, kita bisa melihat siapa yang membeli produk tertentu dan bagaimana pola pembeliannya, sehingga sangat membantu untuk analisis performa produk dan perilaku pelanggan.

In [64]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

Hasil pengelompokan data menunjukkan total penjualan berdasarkan state dan product_type. Setiap wilayah memiliki kontribusi berbeda untuk tiga kategori produk—Jacket, Shirt, dan Trousers. Misalnya, South Australia mencatat jumlah dan nilai penjualan tertinggi untuk kategori Shirt, sementara Queensland unggul pada penjualan Jacket dan Australian Capital Territory menonjol di kategori Trousers. Secara umum, setiap negara bagian memiliki pola permintaan yang bervariasi, tetapi ketiga jenis produk tetap menunjukkan distribusi penjualan yang stabil di seluruh wilayah.

In [65]:
all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
gender            product_type                         
Female            Jacket               481        50963
                  Shirt                445        45146
                  Trousers             481        49295
Male              Jacket               480        50978
                  Shirt                410        41939
                  Trousers             453        46071
Prefer not to say Jacket              2349       251486
                  Shirt               2386       244765
                  Trousers            2400       243092

Data penjualan berdasarkan gender dan jenis produk menunjukkan bahwa pelanggan dengan pilihan gender “Prefer not to say” mendominasi total pembelian di semua kategori—baik Jacket, Shirt, maupun Trousers—dengan jumlah dan nilai penjualan yang jauh lebih tinggi dibandingkan kelompok Female maupun Male. Sementara itu, kelompok Female dan Male memiliki pola pembelian yang relatif seimbang, dengan penjualan yang hanya berbeda tipis antar kategori produk. Secara keseluruhan, preferensi gender tidak terlalu memengaruhi jenis produk yang dibeli, namun jumlah transaksi menunjukkan dominasi kuat dari kelompok yang memilih tidak menyebutkan gender.

In [66]:
all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
age_group product_type                         
Adults    Jacket              2292       245055
          Shirt               2225       227781
          Trousers            2272       231271
Seniors   Jacket               777        82959
          Shirt                733        74977
          Trousers             766        77471
Youth     Jacket               274        29012
          Shirt                301        30842
          Trousers             322        32432

Jika dilihat dari kelompok usia, pelanggan Adults menjadi penyumbang penjualan terbesar untuk semua jenis produk—baik Jacket, Shirt, maupun Trousers—dengan total kuantitas dan nilai transaksi yang jauh lebih tinggi dibandingkan kelompok lainnya. Kelompok Seniors berada di posisi kedua dengan kontribusi yang cukup stabil, sementara Youth memberikan kontribusi paling kecil. Pola ini menunjukkan bahwa kelompok usia produktif (Adults) merupakan pasar utama yang paling aktif berbelanja, sedangkan Youth cenderung membeli dalam jumlah lebih sedikit.

Tugas

Jawablah beberapa pertanyaan berikut ini berdasarkan EDA yang sudah di lakukan

1. Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?
2. Produk apa yang paling banyak dan paling sedikit terjual?
3. Bagaimana demografi pelanggan yang kita miliki?
4. Kapan terakhir pelanggan melakukan transaksi?
5. Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?
6. Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?.

1. Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?

: Secara keseluruhan, performa penjualan stabil dan cukup tinggi. Ini terlihat dari total transaksi yang tersebar merata pada berbagai kategori produk (Jacket, Shirt, Trousers). Revenue yang dihasilkan juga besar, terutama dari produk-produk dengan harga menengah ke atas. Tidak ada tren penurunan signifikan, menunjukkan penjualan berjalan konsisten.

2. Produk apa yang paling banyak dan paling sedikit terjual?

: Paling banyak terjual: Produk dengan jumlah transaksi dan total revenue tertinggi adalah Jacket, Shirt, dan Trousers, dengan penjualan relatif seimbang.

: Jika dilihat per product_name, Denim menjadi yang paling tinggi total revenue-nya, diikuti oleh Casual Slim Fit dan Trench Coat.

: Paling sedikit terjual: Produk seperti Cords dan beberapa pakaian seperti Coach dan Linen memiliki revenue paling rendah.

3. Bagaimana demografi pelanggan yang kita miliki?

Gender: Mayoritas pelanggan memilih kategori “Prefer not to say”, disusul pelanggan Female dan Male.

Usia: Kelompok Adults (25–64 tahun) mendominasi penjualan. Seniors berada di posisi kedua, sedangkan Youth paling sedikit.

Lokasi: Pelanggan tersebar di seluruh wilayah Australia, dengan penjualan tertinggi berasal dari South Australia, Queensland, dan Western Australia.

4. Kapan terakhir pelanggan melakukan transaksi?

: Berdasarkan data order, transaksi pelanggan berlangsung hingga akhir tahun 2021. Jadi aktivitas terakhir mayoritas pelanggan berada pada periode tersebut.

5. Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?

: Frekuensi pembelian pelanggan tergolong rendah–sedang. Umumnya setiap pelanggan hanya melakukan 1–2 transaksi unik, karena jumlah order per customer tidak terlalu besar.

6. Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?.

: Rata-rata pelanggan menghabiskan uang antara $100–$300 per transaksi, tergantung produk dan kuantitas.

: Kelompok Adults menjadi yang paling besar kontribusinya terhadap total revenue perusahaan.

: Pelanggan dalam kelompok Seniors dan Youth menghabiskan lebih sedikit dibandingkan Adults.